installer/bootenv.go

```diff
@@ -91,11 +98,56 @@ func (e *UBootEnv) ReadEnv(names ...string) (BootVars, error) {
 	return vars, err
 }
 
+// ProbeSeparator tries writing 'mender_uboot_separator' to the U-Boot
+// environment using whitespace as a separator. If this write is succesful it
+// means that we are on a legacy U-Boot implementation, which supports the
+// 'key<whitespace>value' syntax. If not, then use the new '=' separator.
+func (e *UBootEnv) ProbeSeparator() (string, error) {
+	log.Info("Probing the Bootloader environment for which separator to use")
+	// Try writing using the old separator syntax
+	err := e.writeEnvImpl(BootVars{
+		menderUBootSeparatorProbe: "1",
+	}, uBootEnvLegacySeparator)
+	if err != nil {
+		return uBootEnvStandardSeparator, err
+	}
+	v, err := e.ReadEnv(menderUBootSeparatorProbe)
+	if err != nil {
+		return uBootEnvStandardSeparator, err
+	}
+	if v[menderUBootSeparatorProbe] == "1" {
+		// Clean variable from the environment
+		err = e.writeEnvImpl(BootVars{
+			menderUBootSeparatorProbe: "",
+		}, uBootEnvLegacySeparator)
+		return uBootEnvLegacySeparator, err
+	}
+	return uBootEnvStandardSeparator, nil
+}
+
+// WriteEnv attempts to write the given 'BootVars' to the bootloader
+// environment.
 func (e *UBootEnv) WriteEnv(vars BootVars) error {
+	// Probe for the separator used by U-Boot. Newer versions support '=',
+	// and libubootenv only supports '=' as the separator, older versions
+	// only support ' '
+	separator, err := e.ProbeSeparator()
+	if err != nil {
+		log.Errorf("Failed to probe the U-Boot environment for which separator to use. Got error: %s", err.Error())
+		return err
+	}
+	log.Debugf("Using (%s) as the bootloader environment separator", separator)
+	return e.writeEnvImpl(vars, separator)
+}
+
+func (e *UBootEnv) writeEnvImpl(vars BootVars, separator string) error {
 	if err := e.checkEnvCanary(); err != nil {
```

Comment:
This check should be kept at the beginning of `WriteEnv`. Not only is it being carried out multiple times here, but it should be before any writing happens to protect the environment.